src/AspNetCore/Src/filewatcher.cxx

```diff
@@ -5,16 +5,45 @@
 
 FILE_WATCHER::FILE_WATCHER() :
     m_hCompletionPort(NULL),
-    m_hChangeNotificationThread(NULL)
+    m_hChangeNotificationThread(NULL),
+    m_fThreadExit(FALSE)
 {
 }
 
 FILE_WATCHER::~FILE_WATCHER()
 {
     if (m_hChangeNotificationThread != NULL)
     {
+        DWORD dwRetryCounter = 20;      // totally wait for 1s
+        DWORD dwExitCode = STILL_ACTIVE;
+
+        // signal the file watch thread to exit
         PostQueuedCompletionStatus(m_hCompletionPort, 0, FILE_WATCHER_SHUTDOWN_KEY, NULL);
-        WaitForSingleObject(m_hChangeNotificationThread, INFINITE);
+        while (!m_fThreadExit && dwRetryCounter > 0)
+        {
+            if (GetExitCodeThread(m_hChangeNotificationThread, &dwExitCode))
+            {
+                if (dwExitCode == STILL_ACTIVE)
+                {
+                    // the file watcher thread will set m_fThreadExit before exit
+                    WaitForSingleObject(m_hChangeNotificationThread, 50);
+                }
```

Comment:
I am pretty sure you need an else here to set m_fThreadExit to true. If the exit code is not active, the thread has already terminated.
The could would be:
```c++
if (dwExitCode == STILL_ACTIVE)
{
    // the file watcher thread will set m_fThreadExit before exit
    WaitForSingleObject(m_hChangeNotificationThread, 50);
}
else
{
    m_fThreadExit = TRUE;
}
```

Reply:
another thread should already set m_fThreadExit to true before its exit

Reply:
Oh I see. That makes sense.